Prepare a guide to local development with Telepresence (#352)
kyma-project/community: guidelines/technical-guidelines/08-telepresence-guide.md

---
title: Using Telepresence for local Kyma development
---

> This document is a general guide to local development with [Telepresence](https://www.telepresence.io/) (tested on version `0.101`).

Reviewer: (tested on version `0.101`). -> It's a common practice in Kyma docs to avoid putting such information in parenthesis. Please incorporate version information into the sentence or make it a note.

> This document is a general guide to local development with [Telepresence](https://www.telepresence.io/) (tested on version `0.101`).

Author: ```suggestion
This document is a general guide to local development with [Telepresence](https://www.telepresence.io/).
```

The Kyma component that you want to debug stores its state in the Kubernetes Custom Resource and, therefore, depends on Kubernetes.    

> Mocking the dependency and developing locally are not possible, and manual deployment on every change is a mundane task.  

Reviewer: ```suggestion
Mocking the dependency and developing locally is not possible, and manual deployment on every change is a mundane task.  
```

Author: I asked about it before and we treat these two things independently, so `are`.

Telepresence is a tool that connects your local process to a remote Kubernetes cluster through proxy, which lets you easily debug locally.  

> When you use Telepresence, it replaces a container in a specified Pod, opens up a new local shell or a pre-configured bash, and proxies the network traffic from the local shell through the Pod. 

Reviewer: I think you can get rid of "When you use Telepresence" and start the sentence with "it replaces".

a specified Pod -> the specified Pod? It's specified, so...

> When you use Telepresence, it replaces a container in a specified Pod, opens up a new local shell or a pre-configured bash, and proxies the network traffic from the local shell through the Pod. 

Reviewer: `opens up a new local shell or a pre-configured bash`
Bash is a shell so why have "a shell OR bash"? 
What is a `pre-configured bash`? I think this could be simplified a bit.

> When you use Telepresence, it replaces a container in a specified Pod, opens up a new local shell or a pre-configured bash, and proxies the network traffic from the local shell through the Pod. 

Author: ```suggestion
It replaces a container in the specified Pod, opens up a new local shell or a pre-configured bash, and proxies the network traffic from the local shell through the Pod. 
```

> Telepresence enables you to make calls such as `curl http://....svc.cluster.local:8081/v1/metadata/services` from your local machine. When you run a server in this shell, other Kubernetes services can access it. 

Reviewer: `calls such as...` - instead of showing a dummy call please describe what kind of calls you can make.

> To start developing with Telepresence, follow these steps:

Author: ```suggestion
To start developing with Telepresence, follow these steps:
> **NOTE**: This guide was tested on version `0.101`.
```

> 1. [Install Telepresence](https://www.telepresence.io/reference/install).

Reviewer: I'm wondering if version information should be included here. Your team tested this on v. 0.101.
Maybe this version should be the minimal required version to use Telepresence to develop Kyma? As in `0.101 or higher`?

2. Run your local Kyma or use the cluster. Then, configure your local kubectl to use the desired Kyma cluster. 

> 3. Check the deployment name to swap, and run: 

Reviewer: I think this should be split into two steps. 
In one step you check the name of the container you want to replace with Telepresence and it's parent deployment, then you run the telepresence magic.

	```
	telepresence --namespace {NAMESPACE} --swap-deployment {DEPLOYMENT_NAME}:{CONTAINER_NAME} --run-shell
	```

> 4. Every Kubernetes Pod has the directory `/var/run/secrets` mounted. The Kubernetes client uses it in the component services. By default, Telepresence copies this directory. It stores the directory path in `$TELEPRESENCE_ROOT`, under the Telepresence shell. The `$TELEPRESENCE_ROOT` variable unwinds to `/tmp/...`. You need to move it to `/var/run/secrets`, where the service expects it. To move it there, create a symlink:

Reviewer: ```suggestion
4. Every Kubernetes Pod has the directory `/var/run/secrets` mounted. The Kubernetes client uses it in the component services. By default, Telepresence copies this directory. It stores the directory path in `$TELEPRESENCE_ROOT`, under the Telepresence shell. The `$TELEPRESENCE_ROOT` variable unwinds to `/tmp/...`. Move it to `/var/run/secrets`, where the service expects it. To move it there, create a symlink:
```

> 4. Every Kubernetes Pod has the directory `/var/run/secrets` mounted. The Kubernetes client uses it in the component services. By default, Telepresence copies this directory. It stores the directory path in `$TELEPRESENCE_ROOT`, under the Telepresence shell. The `$TELEPRESENCE_ROOT` variable unwinds to `/tmp/...`. You need to move it to `/var/run/secrets`, where the service expects it. To move it there, create a symlink:

Author: ```suggestion
4. Every Kubernetes Pod has the directory `/var/run/secrets` mounted. The Kubernetes client uses it in the component services. By default, Telepresence copies this directory. It stores the directory path in `$TELEPRESENCE_ROOT`, under the Telepresence shell. The `$TELEPRESENCE_ROOT` variable unwinds to `/tmp/...`. Move it to `/var/run/secrets`, where the service expects it. To move it there, create a symlink:
```

	```
	sudo ln -s $TELEPRESENCE_ROOT/var/run/secrets /var/run/secrets
	```

> 5. Run `CGO_ENABLED=0 go build ./cmd/{COMPONENT-NAME}` to build the component and give all Kubernetes services that call the component access to this process. The process runs locally on your machine. Use the same command to run various Application Connector services, such as Application Registry, Proxy or Events.

Reviewer: This guide's been component-agnostic so far, keep it that way. Please don't mention any AC components.

> 5. Run `CGO_ENABLED=0 go build ./cmd/{COMPONENT-NAME}` to build the component and give all Kubernetes services that call the component access to this process. The process runs locally on your machine. Use the same command to run various Application Connector services, such as Application Registry, Proxy or Events.

Author: ```suggestion
5. Run `CGO_ENABLED=0 go build ./cmd/{COMPONENT-NAME}` to build the component and give all Kubernetes services that call the component access to this process. The process runs locally on your machine. Use the same command to run various Application Connector services.
```

---
title: Using Telepresence for local Kyma development
---

This document is a general guide to local development with [Telepresence](https://www.telepresence.io/).

The Kyma component that you want to debug stores its state in the Kubernetes Custom Resource and, therefore, depends on Kubernetes.    
Mocking the dependency and developing locally are not possible, and manual deployment on every change is a mundane task.  

Telepresence is a tool that connects your local process to a remote Kubernetes cluster through proxy, which lets you easily debug locally.  
It replaces a container in the specified Pod, opens up a new local shell or a pre-configured bash, and proxies the network traffic from the local shell through the Pod. 

> Telepresence enables you to make calls such as `curl http://....svc.cluster.local:8081/v1/metadata/services` from your local machine. When you run a server in this shell, other Kubernetes services can access it. 

Author: ```suggestion
Telepresence enables you to make HTTP requests from your local machine to services in the cluster that are not exposed outside. When you run a server in this shell, other Kubernetes services can access it. 
```

---
title: Using Telepresence for local Kyma development
---

This document is a general guide to local development with [Telepresence](https://www.telepresence.io/).

Certain Kyma components store their state in Kubernetes custom resources and, therefore, depend on Kubernetes.    
Mocking the dependency and developing locally is not possible, and manual deployment on every change is a mundane task.  

Telepresence is a tool that connects your local process to a remote Kubernetes cluster through proxy, which lets you easily debug locally.  

> It replaces a container in the specified Pod, opens up a new local shell or a pre-configured bash, and proxies the network traffic from the local shell through the Pod. 

Author: ```suggestion
It replaces a container in the specified Pod, opens up a new local shell, and proxies the network traffic from the local shell through the Pod. 
```
